### PR TITLE
Block bindings: Lock editing in fields in editor if meta fields panel is opened.

### LINKED ISSUE #64217: Block bindings: Can't save changes to the source if custom fields are opened
### Description

Since https://github.com/WordPress/gutenberg/pull/61753 it is possible to edit the value of a (post meta) binding in a paragraph block or heading by selecting the block and typing. The change is synced across the site when saved.

But if the user preference to show Custom fields is enabled, meaning the meta boxes are open, the meta box field does not update and using the Save button on the post or page does not save the changes to the field.
After clicking save and refreshing, the value is not updated.

### Step-by-step reproduction instructions

I used the code example from 61753:
1. Register post meta by adding this snippet to your theme's functions.php
```
add_action( 'init', 'test_block_bindings' );

function test_block_bindings() {
	register_meta(
		'post',
		'text_field',
		array(
			'show_in_rest'      => true,
			'single'            => true,
			'type'              => 'string',
			'default'           => 'default text value',
		)
	);
}
```
2. Add a paragraph block bound to the custom field using the Code Editor
```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_field"}}}}} -->
<p>Paragraph content</p>
<!-- /wp:paragraph -->

```
3. Go to Options >Preferences > Advanced and enable custom fields.
4. Select the bond block. Edit the text of the paragraph. Save.
5. Refresh the page.
6. Confirm if the value is updated or not.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

WordPress 6.6.1.

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### PR DESCRIPTION
**Pull request by cbravobernal**
## What?

Fixes #64217
Following @gziolo [proposal](https://github.com/WordPress/gutenberg/issues/64217#issuecomment-2306716812), let's try an experiment that locks editing if the meta fields panel is opened.
 
An alternative to #64505

## Why?

Experimenting.

## Testing Instructions

1) Open a post.
2) Enable custom fields in preferences. Check that the note is added when custom fields are enabled.
3) Bind some connections, check that are not editable.

## Screenshots or screencast 

https://github.com/user-attachments/assets/b203c55f-eae3-46b4-b095-31699dadf0fc

### CODE REVIEW

**Comment by gziolo** [comment]
I like where it is heading. Thank you for opening this alternative PR. We will have to cross-check with https://github.com/WordPress/gutenberg/pull/64570 that adds a new preference for enabling/disabling Block Bindings UI.

**Comment by SantosGuillamot** [comment]
> We will have to cross-check with https://github.com/WordPress/gutenberg/pull/64570 that adds a new preference for enabling/disabling Block Bindings UI.

I see that preference slightly different because they could be compatible. If we follow this approach, what would we want?

* Disable editing of the bound attributes.
* Disable the possibility of creating new connections. We could allow users to connect attributes to custom fields BUT not edit them.

So I guess it is a matter of deciding if we want only the first option or both of them. First option seems enough for me, but I don't have a strong opinion.

**Comment by SantosGuillamot** [comment]
Apart from that, I like the approach. My main concern is that it solves the issue faced in block bindings but other plugins like [this report](https://github.com/WordPress/gutenberg/issues/23078) will still face the same problem when they modify post meta through the store. Additionally, other external sources like ACF, Pods, or WooCommerce will probably need to replicate this logic.

File: packages/edit-post/src/components/preferences-modal/enable-custom-fields.js [hunk 1]
```diff
@@ -61,15 +61,24 @@ export function EnableCustomFieldsOption( { label, areCustomFieldsEnabled } ) {
...
-		<PreferenceBaseOption
-			label={ label }
-			isChecked={ isChecked }
-			onChange={ setIsChecked }
-		>
-			{ isChecked !== areCustomFieldsEnabled && (
-				<CustomFieldsConfirmation willEnable={ isChecked } />
-			) }
-		</PreferenceBaseOption>
+		<>
```
Comment (SantosGuillamot) [inline packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: I believe this extra wrapper is not needed, right?

File: packages/edit-post/src/components/preferences-modal/enable-custom-fields.js [hunk 2]
```diff
@@ -61,15 +61,24 @@ export function EnableCustomFieldsOption( { label, areCustomFieldsEnabled } ) {
...
+			<PreferenceBaseOption
+				label={ label }
+				isChecked={ isChecked }
+				onChange={ setIsChecked }
+			>
+				{ isChecked && (
+					<p className="edit-post-preferences-modal__custom-fields-confirmation-message">
+						{ __(
+							'Connected attributes cannot be edited in the block editor while this interface is active.'
```
Comment (SantosGuillamot) [inline packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: Should this be something like _'Custom fields connected to block attributes cannot be edited in the block editor while this interface is active.'_ or something similar? 

Technically speaking, users aren't editing the attributes but the custom fields.

**Review by SantosGuillamot — COMMENTED** [review]
I left a couple of small comments. Apart from that, the workflow from Unchecked->Checked feels a bit weird to me, although I don't have an idea for improvement right now.
Screenshot 2024-08-30 at 14 59 18

It'd be great to get feedback from @WordPress/gutenberg-design to check what is the proper way of sharing this "warning".

File: packages/edit-post/src/components/preferences-modal/enable-custom-fields.js (hunk 1 above)
Comment (cbravobernal) [reply packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: I tried different approaches and forgot to remove it. 😓

File: packages/edit-post/src/components/preferences-modal/enable-custom-fields.js (hunk 2 above)
Comment (cbravobernal) [reply packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: Done in https://github.com/WordPress/gutenberg/pull/64738/commits/754a17471a9c3077f374f2280b7108584032f932

File: packages/edit-post/src/components/preferences-modal/enable-custom-fields.js (hunk 1 above)
Comment (cbravobernal) [reply packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: Done in https://github.com/WordPress/gutenberg/pull/64738/commits/754a17471a9c3077f374f2280b7108584032f932

File: packages/edit-post/src/components/preferences-modal/enable-custom-fields.js [hunk 3]
```diff
@@ -68,6 +68,13 @@ export function EnableCustomFieldsOption( { label, areCustomFieldsEnabled } ) {
 			isChecked={ isChecked }
 			onChange={ setIsChecked }
 		>
+			{ isChecked && (
+				<p className="edit-post-preferences-modal__custom-fields-confirmation-message">
+					{ __(
+						'Custom fields connected to block attributes cannot be edited in the block editor while this interface is active.'
+					) }
```
Comment (artemiomorales) [inline packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: For me, this wording is still hard to follow. How about a phrasing like this: "Enabling this interface will disable inline editing of custom fields."

The reason I say this: The meta fields panel is still part of the block editor too — at least, It appears that way to an end user. With that in mind, I think specifying 'inline editing' as becoming disabled is clearer.

Additionally, we can likely further simplify the message by removing the 'connected to block attributes' detail. Without the meta fields panel, it's only possible to edit custom fields in the editor when they're connected, right? So to me, specifying extra details around connections just makes the message verbose and difficult to understand.

**Review by artemiomorales — COMMENTED** [review]
This works as expected for me 🙌
I do have a comment on the wording of the message.

File: packages/edit-post/src/components/preferences-modal/enable-custom-fields.js (hunk 3 above)
Comment (gziolo) [reply packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: This is what I see:

Screenshot 2024-09-03 at 12 41 08

The text is so long now that it might be difficult to notice that clicking the reload button is mandatory to activate the setting.

Comment (SantosGuillamot) [reply packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: I share similar concerns regarding the UX [here](https://github.com/WordPress/gutenberg/pull/64738#pullrequestreview-2272127547). I'm in favor of simplifying the message or even looking for an alternative approach (but not sure what).

Comment (gziolo) [reply packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: No change to the message works for me, too. The UI will lock everything as in WordPress 6.5 which will be the clear indicator that it can't be edited inline. One potential enhancement to consider would be showing a prompt to disable Meta Boxes / Custom Fields interface if folks want to use the Attributes in the sidebar. This way, they could reload the page switch between modes. Not the best user experience, but it explains the story well.

Comment (SantosGuillamot) [reply packages/edit-post/src/components/preferences-modal/enable-custom-fields.js]: I'm fine removing the message as well. And we can figure out later if it is needed, or if the Attributes panel is a better place, based on feedback.

**Review by gziolo — APPROVED** [review]
